### PR TITLE
Added basic everest-like test.

### PR DESCRIPTION
**Pull request by joakim-hove**
**Task**
Refcatoring to enable a dynamic number of realisations. See commit message for more extensive explanation.

**Approach**
A very large part of the job has been to downplay the importance of the enkf_state object - which represent a realisation which is supposed to have application lifetime. The enkf_state object still exists, but is *much* less important than previously.

**Pre un-WIP checklist**
- [x] Statoil tests pass locally
- [x] Have completed graphical integration test steps

Statoil/ert#81

### CODE REVIEW

**Comment by markusdregi** [comment]
Could this be moved to the Everest repo? I do not really see why libres should have any notion of Everest..

**Comment by pgdr** [comment]
I agree with Markus.  Res should in my opinion have no knowledge of Everest.

**Comment by pgdr** [comment]
(And if this is a test of `ext_param` then I think we should rename the test accordingly.)

**Comment by joakim-hove** [comment]
Well guys please look at the code before you comment. 

The purpose of this test is to create a test of libres functionality which I *assume* will be required by Everest.

**Comment by pgdr** [comment]
Yes, I read the code, and I still think it should test res functionality, not Everest functionality.

It's WIP so we can wait until it becomes unWIPed before we comment.

**Comment by joakim-hove** [comment]
Ping

**Comment by markusdregi** [comment]
@joakim-hove Is this one ready for review? Tests seems to still be failing..

**Comment by joakim-hove** [comment]
Yrs

File: libenkf/src/ecl_config.c [hunk 1]
```diff
@@ -88,17 +88,20 @@ struct ecl_config_struct
...
  */
 
 bool ecl_config_active(const ecl_config_type * config)
 {
-  if ((config->eclbase == NULL )&& (config->data_file == NULL))
-    return false;
-  else
+  if (config->have_eclbase)
+    return true;
+
```
Comment (markusdregi) [inline libenkf/src/ecl_config.c:97]: How about: `return config->have_eclbase || config->data_file;`?

File: libenkf/src/ecl_config.c [hunk 2]
```diff
@@ -604,27 +587,17 @@ static void handle_has_schedule_file_key(ecl_config_type * ecl_config,
   const char * schedule_target_file = NULL;
 
   if (config_content_node_get_size( schedule_node ) > 1) {
-    schedule_target_file = config_content_node_iget_as_abspath(schedule_node, 1);
```
Comment (markusdregi) [inline libenkf/src/ecl_config.c:607]: Will the docs be updated accordingly?

File: libenkf/src/ecl_config.c [hunk 3]
```diff
@@ -953,6 +920,11 @@ bool ecl_config_get_unified_summary(const ecl_config_type * ecl_config)
...
 }
 
+bool ecl_config_have_eclbase(const ecl_config_type * ecl_config) {
+  return ecl_config->have_eclbase;
+}
+
+
 void ecl_config_add_config_items(config_parser_type * config)
 {
   config_schema_item_type * item;
```
Comment (markusdregi) [inline libenkf/src/ecl_config.c:842]: It we are not supporting to provide a target schedule file any more, please update the definition of SCHEDULE_FILE_KEY accordingly.

File: libenkf/src/ecl_config.c [hunk 4]
```diff
@@ -806,45 +778,45 @@ void ecl_config_free(ecl_config_type * ecl_config)
...
- included. Observe that it is __assumed__ that ecl_util_escape_kw()
- has already been called on the kw.
- */
+   This function checks whether the static kw should be
+   included. Observe that it is __assumed__ that ecl_util_escape_kw()
+   has already been called on the kw.
+*/
 
 bool ecl_config_include_static_kw(const ecl_config_type * ecl_config, const char * kw)
 {
```
Comment (markusdregi) [inline libenkf/src/ecl_config.c]: When you've already touched this, how about:
```
return (ecl_config->include_all_static_kw ||
        set_has_key(ecl_config->static_kw_set, kw) ||
        stringlist_contains(ecl_config->user_static_kw, kw));
```

File: libenkf/src/ecl_config.c [hunk 5]
```diff
@@ -1003,82 +975,6 @@ void ecl_config_add_config_items(config_parser_type * config)
   config_schema_item_set_argc_minmax(item, 1, 1);
 }
 
-void ecl_config_fprintf_config(const ecl_config_type * ecl_config, FILE * stream)
```
Comment (markusdregi) [inline libenkf/src/ecl_config.c:1006]: Happy times! Cool to see that you took the time to remove some code..

File: libenkf/src/enkf_main.c [hunk 6]
```diff
@@ -1335,37 +1337,39 @@ void enkf_main_isubmit_job( enkf_main_type * enkf_main , run_arg_type * run_arg
...
 static void * enkf_main_create_run_path__( enkf_main_type * enkf_main,
                                            const ert_run_context_type * run_context) {
 
-  const bool_vector_type * iactive = ert_run_context_get_iactive(run_context);
-  const int active_ens_size = util_int_min( bool_vector_size( iactive ) , enkf_main_get_ensemble_size( enkf_main ));
   int iens;
-  for (iens = 0; iens < active_ens_size; iens++) {
-    if (bool_vector_iget(iactive , iens)) {
-      run_arg_type * run_arg = ert_run_context_iens_get_arg( run_context , iens);
+  fprintf(stderr,"  [%d,%d)  \n",0, ert_run_context_get_size( run_context ));
```
Comment (markusdregi) [inline libenkf/src/enkf_main.c]: I guess this should be removed.

File: libenkf/src/enkf_main.c [hunk 7]
```diff
@@ -1479,6 +1490,7 @@ static void enkf_main_start_queue(enkf_main_type * enkf_main,
                                   bool verbose_queue) {
 
   job_queue_manager_type * queue_manager = job_queue_manager_alloc( job_queue );
+  fprintf(stderr,"Starting ququq %s \n",__func__);
```
Comment (markusdregi) [inline libenkf/src/enkf_main.c]: Should go?

File: libenkf/src/enkf_state_nodes.c [hunk 8]
```diff
@@ -81,21 +69,18 @@ static void enkf_state_internalize_eclipse_state(enkf_state_type * enkf_state ,
...
+      if (enkf_config_node_get_impl_type(config_node) != FIELD)
         continue;
 
-      if (!enkf_node_has_func(enkf_node , forward_load_func))
+      if (!enkf_config_node_use_forward_init(config_node))
         continue;
 
-      if (internalize_state || enkf_node_internalize(enkf_node , report_step)) {
-
+      if (true) {
```
Comment (markusdregi) [inline libenkf/src/enkf_state_nodes.c]: `&& !false` ? ;) Please remove..

File: libenkf/src/ensemble_config.c [hunk 9]
```diff
@@ -813,22 +818,37 @@ bool ensemble_config_has_impl_type(const  ensemble_config_type * config, const e
...
+        break;
+      }
     }
   }
   hash_iter_free(iter);
   return ret;
 }
 
 
+bool ensemble_config_require_summary(const  ensemble_config_type * ens_config) {
```
Comment (markusdregi) [inline libenkf/src/ensemble_config.c:841]:
```
return  (ensemble_config_has_impl_type(ens_config, SUMMARY) ||
         ensemble_config_GEN_DATA_require_summary(ens_config));
```

File: libenkf/src/forward_load_context.c [hunk 10]
```diff
@@ -54,16 +54,18 @@ UTIL_IS_INSTANCE_FUNCTION( forward_load_context , FORWARD_LOAD_CONTEXT_TYPE_ID)
...
-    const char * eclbase                   = load_context->eclbase;
+    const char * eclbase                   = run_arg_get_job_name( load_context->run_arg );
 
     const bool fmt_file                    = ecl_config_get_formatted(load_context->ecl_config);
     char * header_file                     = ecl_util_alloc_exfilename(run_path , eclbase , ECL_SUMMARY_HEADER_FILE , fmt_file , -1);
     char * unified_file                    = ecl_util_alloc_exfilename(run_path , eclbase , ECL_UNIFIED_SUMMARY_FILE , fmt_file ,  -1);
     stringlist_type * data_files           = stringlist_alloc_new();
 
+
+    printf("Looking for: %s \n",eclbase);
```
Comment (markusdregi) [inline libenkf/src/forward_load_context.c]: Should go?

File: libenkf/src/model_config.c [hunk 11]
```diff
@@ -546,18 +510,30 @@ void model_config_init(model_config_type * model_config ,
...
-    model_config_set_case_table( model_config , ens_size , config_content_iget( config , CASE_TABLE_KEY , 0,0));
-
   if (config_content_has_item( config , ENSPATH_KEY))
     model_config_set_enspath( model_config , config_content_get_value_as_path(config , ENSPATH_KEY));
 
   if (config_content_has_item( config , DATA_ROOT_KEY))
     model_config_set_data_root( model_config , config_content_get_value_as_path(config , DATA_ROOT_KEY));
 
-  if (config_content_has_item( config , JOBNAME_KEY))
+  /*
```
Comment (markusdregi) [inline libenkf/src/model_config.c:519]: Am I understanding it correctly if we could now require `JOBNAME` to be present and then have a boolean flag `ECL` that would toogle the settings in `ecl_config`? I understand that we have legacy, config files that should still work and so forth, just wondering to check that I got it correctly.

Also, will the docs be updated accordingly?

File: libenkf/src/model_config.c [hunk 12]
```diff
@@ -807,10 +769,6 @@ static void model_config_init_user_config(config_parser_type * config ) {
   /*****************************************************************/
   /* Required keywords from the ordinary model_config file */
 
-  item = config_add_schema_item(config, CASE_TABLE_KEY, false);
```
Comment (markusdregi) [inline libenkf/src/model_config.c:810]: Sorry for screaming about the documentation all the time, but...

File: python/python/res/server/simulation_context.py [hunk 13]
```diff
@@ -26,28 +26,29 @@ def __init__(self, ert, sim_fs, mask, itr , verbose=False):
...
+        jobname_fmt = self._ert.getModelConfig().getJobnameFormat()
+        self._run_context = ErtRunContext( EnkfRunType.ENSEMBLE_EXPERIMENT, sim_fs, None, mask, path_fmt, jobname_fmt, subst_list, itr)
 
-        
-    def addSimulation(self, iens):
-        if iens >= len(self._mask):
-            raise UserWarning("Realization number out of range: %d >= %d" % (iens, len(self._mask)))
+
+    def addSimulation(self, iens, geo_id):
+        if iens >= len(self._run_context):
```
Comment (markusdregi) [inline python/python/res/server/simulation_context.py]: Should this be
```python
if not (0 <= iens < len(self._run_context)):
    raise IndexError(..)
```
?

File: python/tests/res/enkf/test_enkf.py [hunk 14]
```diff
@@ -211,31 +211,22 @@ def test_run_context(self):
...
-                run_context.iensGet(1)
-
-            arg0 = run_context[0]
-            arg2 = run_context.iensGet( 2 )
-            #self.assertEqual( arg0 , arg2 )
 
 
     def test_run_context_from_external_folder(self):
-        with TestAreaContext('enkf_test') as work_area:
+        with TestAreaContext('enkf_test', store_area = True) as work_area:
```
Comment (markusdregi) [inline python/tests/res/enkf/test_enkf.py:229]: Should `store_area` be defaulted again perhaps?

File: python/tests/res/enkf/test_everest.py [hunk 15]
```diff
@@ -0,0 +1,90 @@
...
+from ecl.util import BoolVector
+
+from res.test import ErtTestContext
+from res.enkf import EnkfFs, EnkfConfigNode, NodeId, EnkfNode
+from res.enkf import EnKFMain, ErtRunContext
+from res.enkf.enums import EnKFFSType
+from res.enkf.enums import RealizationStateEnum
+
+
+class EverestTest(ExtendedTestCase):
```
Comment (markusdregi) [inline python/tests/res/enkf/test_everest.py]: Can we please rename this test? There is no reason that _libres_ should have the slightest clue what _Everest_ is, and the same goes mentally for us developers when doing work in _libres_. Could we find a name that is descriptive of the new functionality tested instead?

File: python/tests/res/enkf/test_everest.py [hunk 16]
```diff
@@ -0,0 +1,90 @@
...
+    def setUp(self):
+        pass
+
+
+    def test_run(self):
+        ens_size = 2
+        config_file = self.createTestPath("local/config/everest/config.ert")
+        with ErtTestContext( "simple_everest", model_config = config_file, store_area = True) as ctx:
```
Comment (markusdregi) [inline python/tests/res/enkf/test_everest.py]: Let `store_area` default?

File: python/tests/res/enkf/test_everest.py [hunk 17]
```diff
@@ -0,0 +1,90 @@
...
+                state_map[iens] = RealizationStateEnum.STATE_INITIALIZED
+
+            mask = BoolVector( default_value = True, initial_size = batch_size )
+            model_config = ert.getModelConfig( )
+            runpath_fmt = model_config.getRunpathFormat( )
+            jobname_fmt = model_config.getJobnameFormat( )
+            subst_list = ert.getDataKW( )
+            itr = 0
+            run_context = ErtRunContext.ensemble_experiment( sim_fs, mask, runpath_fmt, jobname_fmt, subst_list, itr)
+            print ens_size, batch_size
```
Comment (markusdregi) [inline python/tests/res/enkf/test_everest.py]: Remove

File: python/tests/res/enkf/test_programmatic_res_config.py [hunk 18]
```diff
@@ -337,9 +337,6 @@ def assert_equal_ecl_config(self, loaded_ecl_config, prog_ecl_config):
         self.assertEqual(loaded_ecl_config.get_gridfile(),
                          prog_ecl_config.get_gridfile())
 
-        self.assertEqual(loaded_ecl_config.getEclBase(),
```
Comment (markusdregi) [inline python/tests/res/enkf/test_programmatic_res_config.py:340]: ~~See my comments regarding this in the other PR.~~ Has been resolved as not an issue.

File: test-data/local/config/everest/config.ert [hunk 19]
```diff
@@ -0,0 +1,14 @@
+NUM_REALIZATIONS 2
```
Comment (markusdregi) [inline test-data/local/config/everest/config.ert]: Please rename the `everest` part of the path and remove it from the comment.

File: python/python/res/enkf/ert_run_context.py [hunk 20]
```diff
@@ -49,39 +49,49 @@ def __init__(self , run_type , sim_fs, target_fs , mask , path_fmt , subst_list
...
         run_context._mask = mask
         run_context._path_fmt = path_fmt
         run_context._subst_list = subst_list
-        
+
         return run_context
 
-        
+
+    def is_active(self, index):
```
Comment (markusdregi) [inline python/python/res/enkf/ert_run_context.py:79]: Please add a small test

File: test-data/local/snake_oil_no_data/snake_oil.ert [hunk 21]
```diff
@@ -1,6 +1,6 @@
 QUEUE_SYSTEM LOCAL
+ECLBASE SNAKE_OIL_FIELD
```
Comment (markusdregi) [inline test-data/local/snake_oil_no_data/snake_oil.ert:2]: Could you add a small test that displays that if both `ECLBASE` and `JOBNAME` is present, `JOBNAME` will indeed be used.

**Review by markusdregi — APPROVED** [review]
This is very good work! As I see it, for PR's like this the major work is in realizing what should be done, what is the role of the various objects and what is their responsibilities, how long should they live and so forth. All of this is done very well and I completely agree with these decisions.

However, when this is materialized in code, it is to a large extent about shuffling the right objects to the right places. Which obviously results in large diffs, but it also results in most of my comments regarding the actual code being nitpicking. TLDR; Do not use the comments in the actual code to extrapolate my general opinion regarding this work, because that is highly unfair to this PR. 🥇

**Comment by markusdregi** [comment]
The tests are currently failing due to Statoil/libecl#232 not compiling. However, that boils down to removing a single unused variable..

File: libenkf/src/model_config.c (hunk 12 above)
Comment (joakim-hove) [reply libenkf/src/model_config.c:810]: The `CASE_TABLE` keyword has been removed from the documentation.

File: test-data/local/snake_oil_no_data/snake_oil.ert (hunk 21 above)
Comment (joakim-hove) [reply test-data/local/snake_oil_no_data/snake_oil.ert:2]: Done

File: python/python/res/enkf/ert_run_context.py (hunk 20 above)
Comment (joakim-hove) [reply python/python/res/enkf/ert_run_context.py:79]: Done

File: python/tests/res/enkf/test_everest.py (hunk 15 above)
Comment (joakim-hove) [reply python/tests/res/enkf/test_everest.py]: Done

File: python/tests/res/enkf/test_everest.py (hunk 16 above)
Comment (joakim-hove) [reply python/tests/res/enkf/test_everest.py]: Done

File: python/tests/res/enkf/test_everest.py (hunk 17 above)
Comment (joakim-hove) [reply python/tests/res/enkf/test_everest.py]: Done

File: test-data/local/config/everest/config.ert (hunk 19 above)
Comment (joakim-hove) [reply test-data/local/config/everest/config.ert]: Done

File: python/tests/res/enkf/test_enkf.py (hunk 14 above)
Comment (joakim-hove) [reply python/tests/res/enkf/test_enkf.py:229]: Yes

**Comment by joakim-hove** [comment]
Thank you for thorough review - I think all comments should be adressed now.

**Comment by markusdregi** [comment]
Very good 👍 You have a minor syntax error in Python. Thank you for addressing the comments

**Comment by joakim-hove** [comment]
OK - this has been approved, and is currently green - I'll merge it. Mind you this is big - touching the golden triangle of important, untested and old code - i.e. fallouts will not surprise me very much.

**Pull request merged**